### PR TITLE
adds kiota to dotnet openapi tool

### PR DESCRIPTION
**Pull request by baywet**
follow up of https://github.com/baywet/aspnetcore/pull/1 CC @dougbu

## Questions left:

- Should the dependencies be added by the api definition package or by this process (through the same mechanism that adds the ApiDescription package)?
- How are the TypeScript dependencies supposed to be added since we're talking about a different package manager?
- Would it be possible to add target namespace, target folder, and api client class name parameters to the tool?
- Why is the Option for CodeGenerator not using a generic type to pass the enum and leave the parsing to system.command line?

## Resulting CSProj (for now):

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.OpenApi.Kiota.ApiDescription.Client" Version="0.5.0-preview" />
  </ItemGroup>
  <ItemGroup>
    <OpenApiReference Include="Mail.yml" SourceUrl="https://raw.githubusercontent.com/microsoftgraph/msgraph-sdk-powershell/dev/openApiDocs/v1.0/Mail.yml" CodeGenerator="KiotaCSharp" />
  </ItemGroup>
</Project>
``` 

### Create a local testing project with a nuget.config

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
	<packageSources>
		<add key="localnuget" value="file:///workspaces/localnuget" />
	</packageSources>
</configuration>
```

### Add Kiota and Kiota ApiDescription to the local source

dotnet pack for kiota and kiota api ddescription, copy in the previous local source folder

### Pack the tool and run it

```shell
cd src/Tools/Microsoft.dotnet-openapi/src
dotnet pack
dotnet tool install --local Microsoft.dotnet-openapi --add-source /workspaces/aspnetcore/artifacts/packages/Debug/Shipping/ --version 7.0.0-dev
dotnet tool run dotnet-openapi -- add url https://raw.githubusercontent.com/microsoftgraph/msgraph-sdk-powershell/dev/openApiDocs/v1.0/Mail.yml -c KiotaCSharp -p /workspaces/dummy/dummy.csproj
dotnet tool uninstall --local Microsoft.dotnet-openapi
```

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=41928) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=41928) before we can accept your contribution.<br/><br/>:x: baywet [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=41928)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=41928) it.</sub>

File: src/Tools/Microsoft.dotnet-openapi/src/Microsoft.dotnet-openapi.csproj [hunk 1]
```diff
@@ -31,5 +31,10 @@
       <_Parameter2>$(NewtonsoftJsonVersion)</_Parameter2>
       <_Parameter3>NSwagCSharp;NSwagTypeScript</_Parameter3>
     </AssemblyAttribute>
+    <AssemblyAttribute Include="Microsoft.DotNet.Openapi.Tools.Internal.OpenApiDependencyAttribute">
+      <_Parameter1>Microsoft.OpenApi.Kiota.ApiDescription.Client</_Parameter1>
```
Comment (dougbu) [inline src/Tools/Microsoft.dotnet-openapi/src/Microsoft.dotnet-openapi.csproj:35]: Referencing only this package, will updated projects using `KiotaCSharp` or `KiotaTypeScript` be able to successfully use a generated client❔

Comment (baywet) [reply src/Tools/Microsoft.dotnet-openapi/src/Microsoft.dotnet-openapi.csproj:35]: what would be the recommended approach from your point of view? add those dependencies to the APIDescription package?
Or add those dependencies to the target through the same mechanism we're adding the API description package?
I prefer the later as it provides more control to users over which dependencies and versions dependencies they bring to their project should they want to change things.

Also, how are the JS/TS/npm packages added? are they being wrapped in a nuget package? or is there a possibility to add a reference to an npm package? if so how?

**Comment by dougbu** [comment]
> Would it be possible to add target namespace, target folder, and api client class name parameters to the tool?

That sounds like additional customization users could do based on the incredibly poor "documentation" in <https://github.com/dotnet/aspnetcore/blob/main/src/Tools/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.props>. I would recommend adjusting the defaults to your preferences in the Microsoft.OpenApi.Kiota.ApiDescription.Client package and adding a bit more information for users over complicating the tool w/ lots of new options. Editing the updated project file isn't that big a deal.

> Why is the Option for CodeGenerator not using a generic type to pass the enum and leave the parsing to system.command line?

I don't remember the full history. @captainsafia made the most recent changes to the tool, @ryanbrandenburg wrote it, and @BillHiebert did the VS side of things and designed the service that provides some of the versioning. @bradygaster may also have context.

File: src/Tools/Microsoft.dotnet-openapi/src/Microsoft.dotnet-openapi.csproj (hunk 1 above)
Comment (baywet) [reply src/Tools/Microsoft.dotnet-openapi/src/Microsoft.dotnet-openapi.csproj:35]: added the dependencies for dotnet. waiting on @bradygaster 's input for the typescript ones

**Comment by baywet** [comment]
thanks for the additional information. I'm not able to understand how this file works and what pieces I need just from reading it. Schedule a meeting today with you to go over that together.

**Comment by baywet** [comment]
Update for everyone watching this PR, during our connect with Doug we've made some updates to the kiota packages that I still need to test and publish. https://github.com/microsoft/kiota/pull/1604
Other than that, I need to connect with Brady to learn more about the TypeScript dependencies.

**Comment by baywet** [comment]
Update, I've made progress thanks to the help provided by Doug, and answer provided by Vijay

## TypeScript

There are effectively two cases:
- asp.net mvc app (no SPA), the TypeScript source file gets generated to the obj folder. It's expected that people will copy it/serve it so they can use it on the front end. NSwag only relies on dependencies that are already in the Visual Studio TypeScript targets. So here we might be at a dead end where the build will complain about missing dependencies and we don't have a way to add them (either manually or programmatically)
- asp.net with SPA included. In this case a client SPA gets scaffolded with a package.json in ClientApp directory (or equivalent, denominated by SPARoot). In that case we try to run an npm install command, and if it fails, display a comprehensive message.

## Dotnet

I got to a place where the packaging and the command seems to be structured properly, if I execute the command manually in the terminal, it works. However when I try to build the project (dotnet build) I get the following error message. I'm not sure where to go from there.

```
MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
Microsoft.Build.Framework.InternalErrorException: MSB0001: Internal MSBuild Error: Unexpected CopyLocal flag.
at Microsoft.Build.Tasks.CopyLocalStateUtility.IsCopyLocal(CopyLocalState state)
at Microsoft.Build.Tasks.ResolveAssemblyReference.LogResults(ReferenceTable dependencyTable, List`1 idealAssemblyRemappings, List`1 idealAssemblyRemappingsIdentities, List`1 generalResolutionExceptions)
at Microsoft.Build.Tasks.ResolveAssemblyReference.Execute(FileExists fileExists, DirectoryExists directoryExists, GetDirectories getDirectories, GetAssemblyName getAssemblyName, GetAssemblyMetadata getAssemblyMetadata, GetLastWriteTime getLastWriteTime, GetAssemblyRuntimeVersion getRuntimeVersion, GetAssemblyPathInGac getAssemblyPathInGac, IsWinMDFile isWinMDFile, ReadMachineTypeFromPEHeader readMachineTypeFromPEHeader)
at Microsoft.Build.Tasks.ResolveAssemblyReference.Execute()
at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
at Microsoft.Build.BackEnd.TaskBuilder.InitializeAndExecuteTask(TaskLoggingContext taskLoggingContext, ItemBucket bucket, IDictionary`2 taskIdentityParameters, TaskHost taskHost, TaskExecutionMode howToExecuteTask)
at Microsoft.Build.BackEnd.TaskBuilder.ExecuteBucket(TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask, Dictionary`2 lookupHash)
at Microsoft.Build.BackEnd.TaskBuilder.ExecuteTask(TaskExecutionMode mode, Lookup lookup)
at Microsoft.Build.BackEnd.TaskBuilder.ExecuteTask(TargetLoggingContext loggingContext, BuildRequestEntry requestEntry, ITargetBuilderCallback targetBuilderCallback, ProjectTargetInstanceChild taskInstance, TaskExecutionMode mode, Lookup inferLookup, Lookup executeLookup, CancellationToken cancellationToken)
at Microsoft.Build.BackEnd.TargetEntry.ProcessBucket(ITaskBuilder taskBuilder, TargetLoggingContext targetLoggingContext, TaskExecutionMode mode, Lookup lookupForInference, Lookup lookupForExecution)
at Microsoft.Build.BackEnd.TargetEntry.ExecuteTarget(ITaskBuilder taskBuilder, BuildRequestEntry requestEntry, ProjectLoggingContext projectLoggingContext, CancellationToken cancellationToken)
at Microsoft.Build.BackEnd.TargetBuilder.ProcessTargetStack(ITaskBuilder taskBuilder)
at Microsoft.Build.BackEnd.TargetBuilder.BuildTargets(ProjectLoggingContext loggingContext, BuildRequestEntry entry, IRequestBuilderCallback callback, String[] targetNames, Lookup baseLookup, CancellationToken cancellationToken)
at Microsoft.Build.BackEnd.RequestBuilder.BuildProject()
at Microsoft.Build.BackEnd.RequestBuilder.BuildAndReport()
at Microsoft.Build.BackEnd.RequestBuilder.RequestThreadProc(Boolean setThreadParameters)
```

**Comment by adityamandaleeka** [comment]
I just noticed this PR. Looks like it's been silent for nearly a year... is this still being worked on or should it be closed at this point?

**Comment by captainsafia** [comment]
I think it should be fine to close this. We'll likely want to put in some design work into how the client gen code can support new providers. @baywet Thoughts?

**Comment by baywet** [comment]
Thanks for the nudge here. Deferring to @darrelmiller who wanted me to keep things open until discussions pan out a while ago.

**Pull request closed without merging**